File: Tests/PersistenceTests/FileDrawingRepositoryTests.swift
import Foundation
import Testing
@testable import Scratchpad

struct FileDrawingRepositoryTests {

    private func makeRepository() -> (FileDrawingRepository, URL) {
        let directory = FileManager.default.temporaryDirectory
            .appendingPathComponent("ScratchpadTests-\(UUID().uuidString)", isDirectory: true)
        return (FileDrawingRepository(directory: directory), directory)
    }

    @Test func loadingMissingDisplayReturnsEmptyDocument() {
        let (repository, directory) = makeRepository()
        defer { try? FileManager.default.removeItem(at: directory) }

        let document = repository.load(for: DisplayIdentifier(rawValue: 1))
        #expect(document == .empty)
    }

    @Test func savedDocumentRoundTripsThroughLoad() throws {
        let (repository, directory) = makeRepository()
        defer { try? FileManager.default.removeItem(at: directory) }

        let display = DisplayIdentifier(rawValue: 42)
        let stroke = Stroke(points: [StrokePoint(x: 0, y: 0), StrokePoint(x: 10, y: 10)])
        let document = DrawingDocument(strokes: [stroke], updatedAt: Date(timeIntervalSince1970: 0))

        try repository.save(document, for: display)
        let loaded = repository.load(for: display)

        #expect(loaded == document)
    }

    @Test func clearRemovesTheSavedDocument() throws {
        let (repository, directory) = makeRepository()
        defer { try? FileManager.default.removeItem(at: directory) }

        let display = DisplayIdentifier(rawValue: 7)
        try repository.save(DrawingDocument(strokes: [Stroke()], updatedAt: Date()), for: display)

        try repository.clear(for: display)

        #expect(repository.load(for: display) == .empty)
    }

    @Test func differentDisplaysArePersistedIndependently() throws {
        let (repository, directory) = makeRepository()
        defer { try? FileManager.default.removeItem(at: directory) }

        let first = DisplayIdentifier(rawValue: 1)
        let second = DisplayIdentifier(rawValue: 2)
        let firstDocument = DrawingDocument(strokes: [Stroke()], updatedAt: Date())

        try repository.save(firstDocument, for: first)

        #expect(repository.load(for: first) == firstDocument)
        #expect(repository.load(for: second) == .empty)
    }
}

File: Windows/Scratchpad.Tests/PersistenceTests/FileDrawingRepositoryTests.cs
using Scratchpad.Drawing;
using Scratchpad.Overlay;
using Scratchpad.Persistence;
using Xunit;

namespace Scratchpad.Tests.PersistenceTests;

public class FileDrawingRepositoryTests
{
    private static (FileDrawingRepository Repository, string Directory) MakeRepository()
    {
        var directory = Path.Combine(Path.GetTempPath(), $"ScratchpadTests-{Guid.NewGuid()}");
        return (new FileDrawingRepository(directory), directory);
    }

    [Fact]
    public void LoadingMissingDisplayReturnsEmptyDocument()
    {
        var (repository, directory) = MakeRepository();
        try
        {
            var document = repository.Load(new DisplayIdentifier("\\.\\DISPLAY1"));
            Assert.Empty(document.Strokes);
        }
        finally
        {
            if (Directory.Exists(directory)) Directory.Delete(directory, recursive: true);
        }
    }

    [Fact]
    public void SavedDocumentRoundTripsThroughLoad()
    {
        var (repository, directory) = MakeRepository();
        try
        {
            var display = new DisplayIdentifier("\\.\\DISPLAY42");
            var stroke = new Stroke { Points = { new StrokePoint(0, 0), new StrokePoint(10, 10) } };
            var document = new DrawingDocument { Strokes = { stroke } };

            repository.Save(document, display);
            var loaded = repository.Load(display);

            Assert.Single(loaded.Strokes);
            Assert.Equal(stroke.Points, loaded.Strokes[0].Points);
        }
        finally
        {
            if (Directory.Exists(directory)) Directory.Delete(directory, recursive: true);
        }
    }

    [Fact]
    public void ClearRemovesTheSavedDocument()
    {
        var (repository, directory) = MakeRepository();
        try
        {
            var display = new DisplayIdentifier("\\.\\DISPLAY7");
            repository.Save(new DrawingDocument { Strokes = { new Stroke() } }, display);

            repository.Clear(display);

            Assert.Empty(repository.Load(display).Strokes);
        }
        finally
        {
            if (Directory.Exists(directory)) Directory.Delete(directory, recursive: true);
        }
    }

    [Fact]
    public void DifferentDisplaysArePersistedIndependently()
    {
        var (repository, directory) = MakeRepository();
        try
        {
            var first = new DisplayIdentifier("\\.\\DISPLAY1");
            var second = new DisplayIdentifier("\\.\\DISPLAY2");
            var firstDocument = new DrawingDocument { Strokes = { new Stroke() } };

            repository.Save(firstDocument, first);

            Assert.Single(repository.Load(first).Strokes);
            Assert.Empty(repository.Load(second).Strokes);
        }
        finally
        {
            if (Directory.Exists(directory)) Directory.Delete(directory, recursive: true);
        }
    }
}
